posts/2022-06-27-foldleft-introduction.adoc

```diff
@@ -0,0 +1,277 @@
+Intéressons-nous tout d'abord à la signature de cette méthode
+[source, scala]
+----
+def foldLeft[B](z: B)(op: (B, A) => B): B
+----
+Nous voyons ici plusieurs choses intéressantes :
+1. La fonction est typée B, qui est le type de sortie
+2. Elle prend en entrée un premier paramètre z (zéro) de ce fameux type B. Ce paramètre est important car c'est le point de départ de la fonction, il sert à l'initialiser et à collecter les résultats. Ce sera notre accumulateur
+3. Et en second paramètre "op" (opération) une fonction qui prend deux paramètres (B, le type de sortie et A , le type des éléments de la liste) et qui permet d'obtenir ce fameux type de sortie B.
+En gros, cette fonction va parcourir tous les éléments de la liste [A] et va collecter le résultat dans l'accumulateur [B]. A la fin, on se retrouve avec [B].
+Commençons par un exemple simple. Nous allons additionner chaque chiffre d'une liste avec son prédécesseur pour obtenir le résultat final.
+
+image::foldleft-decomposition.png[FoldLeft decomposition, width = 700]
+
+[source, scala]
+----
+val numbers = List(1, 2, 3, 4)
+numbers.foldLeft[Int](z = 0)(op = (accumulator, number) => accumulator + number)
+// Ecriture simplifiée
+numbers.foldLeft(0)(_ + _)
+// Result => Out: Int = 10
+----
+En réalité, nous pouvons utiliser ici la méthode reduce qui n'utilise pas de condition initiale et retourne le même type qu'en entrée.
+[source, scala]
+----
+val numbers = List(1, 2, 3, 4)
+numbers.reduce(_ + _)
+// Ce qui equivaut à
+numbers.sum
+// Result => Out: Int = 10
+----
+Rapprochons-nous petit à petit de cas plus concrets.
+Dans l'exemple suivant, je voudrais, non pas faire la somme des chiffres de la liste, mais obtenir une nouvelle liste dont chaque chiffre serait l'addition du chiffre de la liste "numbers" avec son prédécesseur.
+Nous allons parcourir la liste "List(1, 2, 3, 4)" et à chaque itération, faire cette opération :
+1. accumulateur vide et chiffre courant = 1 : résultat ==> accumulateur = List(1)
+2. accumulateur (1) et chiffre courant = 2 : résultat ==> accumulateur = List(1, 2+1)
+3. accumulateur (1, 3) et chiffre courant = 3 : résultat ==> accumulateur = List(1, 3, 3+3)
+4. accumulateur (1, 3, 6) et chiffre courant = 4 : résultat ==> accumulateur = List(1, 3, 6, 4+6)
+Résultat final : `newList = List(1, 3, 6, 10)``
+Une écriture possible peut s'appuyer sur un pattern matching avec une condition dédiée à la condition initiale, c'est-à-dire lorsque nous démarrons avec une liste vide car nous aurions une erreur à l'exécution si nous appelions `accumulator.last` sur une liste vide. Comme nous n'avons rien à additionner au début, nous remplissons simplement la liste avec le premier chiffre.
+[source, scala]
+----
+val numbers = List(1, 2, 3, 4)
+val newList: Seq[Int] = numbers.foldLeft[Seq[Int]](Seq.empty) {
+    case (accumulator, number) if accumulator.isEmpty => accumulator :+ number
+    case (accumulator, number) => accumulator :+ (accumulator.last + number)
+}
```

Comment:
Guillaume, I would add a remark here about performance/inefficiency of the shown implementation.

`foldLeft[Seq[Int]](Seq.empty) ...` is generic in the sense that it takes a `Seq[Int]`. As you append items to the accumulator (initialised as `Seq.empty[Int]`), this is very inefficient as the compiler will pick `List` as the collection type for the accumulator. The inefficiency presents itself twice in each step of processing the elements in the input sequence: first in getting the last element in the sequence (`accumulator.last`) and secondly when appending the new element at the end of the sequence (`accumulator :+ ...`).
I think you should point this out in the post and you can refer to the [Scala Collection Performance Characteristics](https://docs.scala-lang.org/overviews/collections-2.13/performance-characteristics.html).
I think doing so gives you the possibility to give an alternative solution that, while staying generic, avoids taking the performance hit. It would look like this:
```
List(1,2,3,4).foldLeft(Seq.empty[Int]) {
  case (Nil, el) => Seq(el)
  case (accumulator, el) => (accumulator.head + el) +: accumulator
}
```

In this case, getting the `head` element and prepending a new element to the accumulator are very fast (constant time) operations. Of course, this implementation doesn't produce the same result as the original implementation as the element order is reversed. This can be solved either by using a `foldRight` version of the implementation or calling `reverse` on the `foldLeft` based implementation.

It is interesting to point out that `foldRight` in the Scala Collection library is `foldLeft` in disguise:

From [scala/collection/IterableOnce.scala](https://github.com/scala/scala/blob/v2.13.8/src/library/scala/collection/IterableOnce.scala#L309):

```
def foldRight[B](z: B)(op: (A, B) => B): B = reversed.foldLeft(z)((b, a) => op(a, b))
```

Reply:
Thank you Eric for this very interesting point 👍  
As I'm not very comfortable with this topic (especially to popularise it), would you mind if I paraphrased your comment in a section crediting you? I would also add your link as a source

Reply:
Guillaume, yes, feel free to add it to the blog article (and no need to mention me btw) !